src/dlsnode/storage/BucketFile.d

```diff
@@ -440,6 +570,90 @@ public class BucketFile: OutputStream
 
     /**************************************************************************
 
+        Reads data from the file, not blocking the fiber if there's not enough
+        data, giving the opportunity to the user to suspend in the most convenient
+        location. After the context has been resumed by AsyncIO, the call should
+        be repeated, which will then be guaranteed to succeed or to fail as a
+        result of feof.
+
+        Params:
+            waiting_context = ContextAwaitingJob used to wake
+                              the calling context when the read is completed
+            num_bytes = requested amount of data to read from the file
+            promise = filled with the promise that this instance is doing the
```

Comment:
Am I understanding this correctly that the promise has two roles:
1. On the first call, it's returned in state "made".
2. On the subsequent call, the user is expected to pass it in via this pointer argument. It is then returned again, now fulfilled.
?

Is there ever a reason, in case 1, where you'd also want to pass in a promise as an argument?

Reply:
I can't believe I'm asking this, but should there be some kind of `const`ness involved here? :grimacing: The caller shouldn't be allowed to modify the returned promise, right?

Reply:
Yeah, the only thing that the caller can change is when they "reap" the promise, collecting the result. I'll update.

Reply:
> Is there ever a reason, in case 1, where you'd also want to pass in a promise as an argument?
  
No, I can't think of any reason for this. Probably this can/should be asserted.

Reply:
I was wondering about 2 overloaded methods with different APIs.

Reply:
Right. The advantage of this (IMHO) is that user doesn't need to check if the promise is made _before_ the call, but only after the call: if the promise has been made, we suspend and next time we do exactly the same calling sequence, this time to get the result.

Reply:
Could you give a usage example here to show what you mean?

Reply:
Oh, sorry, I missed this comment. So, this is how it looks right now:

```
    class Reader
    {
        Promise* read_promise = null;

        void foo()
        {
                    end_of_bucket = this.file.nextRecord(waiting_context,
                        this.record_header, read_promise);

                    // Try again when reading is done
                    if (read_promise.promise_made)
                    {
                        return;
                    }
                    // do something with the result
          }
```

So, here `foo()` uses the promise set to `null`. When the first read is promised, `read_promise` will be set, `foo` checks that the promise has been made and exit out. When the caller of foo() is resumed, they don't need to do anything with the promise - they just need to repeat `foo()` call and the existing promise will be used and the data will be read this time, "reaping" the promise (deinitializing it).

Otherwise, if we go with the separate function, `foo()` needs to look like this:


```
    class Reader
    {
        Promise* read_promise = null;

        void foo()
        {
            if (read_promise is null)
            {
                read_promise = this.file.nextRecord(waiting_context,
                    this.record_header);
                // Try again when reading is done
                if (read_promise.promise_made)
                {
                    return;
                 }
             }
             else
             {
                 this.file.nextRecord(waiting_context, this.record_header, read_promise);
                 assert(!read_promise.promise_made); // must be fulfilled now.
              }

              // do something with the result
              this.read_promise = null;
          }
```

Reply:
I still don't quite get it. So you'd call `foo(); foo();`?

Reply:
It would be more like:

```
do {
    if (!foo()) { suspend; continue; }
    resultAvailable = true;
} while (!resultAvailable);
```

Because the request would need to exit `foo()` to go back to the `RequestOnConn` implementation to suspend there. And then it just retries the same operation when resumed (so, it calls `foo` again, but this time `foo` has data for it .

Reply:
Real-life example: https://github.com/sociomantic-tsunami/dlsnode/pull/8/files#diff-f3edcc60e5fec41b6c4012d736c49f04R237

Reply:
I don't really have enough insight into this whole system to comment further on this :/ It _seems_ kinda weird to me, but maybe it's fine? :D Let's leave it for now, if you're happy with it how it is.

Reply:
I have one more approach in my head, let's see if that makes it better. The idea is that I go with the standard C++ route of using promises/futures: http://en.cppreference.com/w/cpp/thread/promise http://en.cppreference.com/w/cpp/thread/future

This is how it would look then:

- Buffer that user reads data from no longer accepts promise as an argument. Instead, it hold the promise internally and _always_ returns the future:

```
class Buffer
{
    // Promise object. It's stored only at the producer's side and will provide
    // future to pass back to the consumer's side
    Promise!(void[]) promise;

    Future!(void[]) read ()
    {
        if (not_enough_data)
        {
             // pass the promise to AIO to fill it up
             this.scheduleAsyncRead(promise);
             return promise.getFuture();
        }
        else
        {
            // reads directly into the future
            this.readImmediately(promise.getFuture());
            return promise.getFuture();
        }
    }
}
```

Now, this is how the caller would look like:

```
class Iterator
{
    Future!(void[]) future; // Future returned from the Buffer
    void nextRecord ()
    {        
        // check if the future is valid (if we suspended and read has finished)
        if (this.future.valid())
        {
            // automatically sets f.valid to false (see: http://en.cppreference.com/w/cpp/thread/future)
            sendToClient(this.future.get()); 
        }
        else
        {
            // There was no read performed, let's read some data
            this.future = this.buffer.read();

           // did we manage to read data without scheduling new AIO (for example,
           // if there was data in the buffer)
           if (this.future.valid())
           {
               sendToClient(f.get());
           }
           else
           {
               // return to RequestOnConn and suspend. When AIO resume this request
               // this.future.valid() will be true and the result will be there.
               return;
           }
        }
    }
}
```

The advantages:

- Consumer doesn't need to call producer twice - instead it stores the future.
- There's clear ownership of everything involved - producer owns promise, consumer stores the future (result) tied to that promise, there are no passing by pointer and by reference, the API stays clean.

We could even go a step further and replace entire `JobNotification` with the promise/future mechanism - instead of passing the `JobNotification` from the consumer to the AIO (the producer), we will return `future` from the producer to the consumer, and the consumer will be able at any point to suspend itself and call `future.wait`, waiting to get suspended. I would do that later, but it seems like a good step towards standardizing AIO subsystem.